### ISSUE TITLE
Non-deterministic overflow in `rustc_thread_pool` causing CI failures

### ISSUE THREAD

**Issue by hkratz** [rust-lang/rust#90227]
It seems that enabling overflow checks (#89776) uncovered non-deterministic overflows in rustc-rayon-core causing the build failures of #90042 and #90222:
* https://github.com/rust-lang-ci/rust/runs/3987259810?check_suite_focus=true#step:25:26550
* https://github.com/rust-lang-ci/rust/runs/3987471878?check_suite_focus=true#step:25:21315

Overflow location:
https://github.com/rust-lang/rustc-rayon/blob/c8ec88d8a2236d1fe19d65a4ab38834f76d256b7/rayon-core/src/sleep/mod.rs#L330

This code is not in upstream rayon but just in our fork, introduced here:
https://github.com/rust-lang/rustc-rayon/commit/27911f7176419b0fa8ca8cc666d44b212077302e#diff-cde9d726ca2f32c319420be6c61d2e57ad9daff7f85a5f8f7be25474a15f9c4dR328

**Comment by the8472** [comment]
> Hm... I would suggest that rustdoc probably wants to use mainline rayon here, rather than rustc's fork of it, since I/O code likely _doesn't_ want to interop with jobserver and such. I don't know whether that will fix this particular issue, but it seems like a good idea.

Is that really the right solution? It seems like the overflow checks have found a bug in `rustc_rayon` which was exposed by the use in rustdoc. Removing the use in rustdoc does not remove the fix additions in the custom rayon version.

Isn't the goal of adding overflow checks to uncover latent bugs and then fix them rather than finding new ways to ignore them?

**Comment by Mark-Simulacrum** [comment]
We should still try to fix the bug, absolutely, and investigate whether it manifests in upstream rayon as well. However, I don't think that changes the assessment that relying on upstream rayon is likely better than the fork in the long run for rustdoc's use case.

**Comment by camelid** [comment]
> Is that really the right solution? It seems like the overflow checks have found a bug in `rustc_rayon` which was exposed by the use in rustdoc. Removing the use in rustdoc does not remove the fix additions in the custom rayon version.

Yeah, we switched to mainline rayon mostly for other reasons. That's why I left this issue open :)

**Comment by saethlin** [comment]
@hkratz Please paste the contents of links in issues instead of just providing a URL. Those links to CI logs now dead, so the only hint now about how to reproduce this is
> on Windows, in rustdoc

**Comment by petrochenkov** [comment]
I'm going to close parallel frontend issues without reproduction reported before 2024, a lot of things changes since then, and no other similar issues were reported.
Please reopen if it ever happens again.

**Comment by cuviper** [comment]
It's not happening because we've hacked around it -- or at least, we can't know if it's still an issue until we remove this:
https://github.com/rust-lang/rust/blob/d2f24127d97d8c795d1cf5fbf6037f9885f34d39/Cargo.toml#L62-L67

**Comment by petrochenkov** [comment]
Reopening, this only can be closed if https://github.com/rust-lang/rust/pull/161296 is not reverted in a month or so.
